libs/MeshKernel/src/Polygons.cpp

```diff
@@ -163,88 +167,61 @@ std::vector<meshkernel::Point> Polygons::RefineFirstPolygon(size_t startIndex,
+        throw std::invalid_argument("Polygons::RefineFirstPolygon: The indices are not valid: " + std::to_string(startIndex) + ", " + std::to_string(endIndex));
     }
 
+    //--------------------------------
+
+    const auto& [outerStart, outerEnd] = m_outer_polygons_indices[polygonIndex];
+
     const auto edgeLengths = PolygonEdgeLengths(m_nodes);
     std::vector<double> nodeLengthCoordinate(edgeLengths.size());
     nodeLengthCoordinate[0] = 0.0;
+
     for (size_t i = 1; i < edgeLengths.size(); ++i)
     {
         nodeLengthCoordinate[i] = nodeLengthCoordinate[i - 1] + edgeLengths[i - 1];
     }
 
-    const auto numNodesRefinedPart = size_t(std::ceil((nodeLengthCoordinate[endIndex] - nodeLengthCoordinate[startIndex]) / refinementDistance) + (double(endIndex) - double(startIndex)));
-    const auto& [outerStart, outerEnd] = m_outer_polygons_indices[polygonIndex];
-    const auto numNodesNotRefinedPart = startIndex - outerStart + outerEnd - endIndex;
-    const auto totalNumNodes = numNodesRefinedPart + numNodesNotRefinedPart;
+    // Approximate number of nodes in the refined sections.
+    const size_t numNodesRefinedPart = static_cast<size_t>(std::ceil((nodeLengthCoordinate[endIndex] - nodeLengthCoordinate[startIndex]) / refinementDistance)) + endIndex - startIndex;
+    size_t numNodesNotRefinedPart = startIndex - outerStart + outerEnd - endIndex;
+    // Approximate the number of nodes in the refined polygon.
+    size_t totalNumNodes = numNodesRefinedPart + numNodesNotRefinedPart;
     std::vector<Point> refinedPolygon;
+
     refinedPolygon.reserve(totalNumNodes);
 
-    // before refinement
-    for (auto i = outerStart; i <= startIndex; ++i)
+    // Add nodes before the section to be refined
+    for (size_t i = outerStart; i <= startIndex; ++i)
     {
         refinedPolygon.emplace_back(m_nodes[i]);
     }
 
-    // refined part
-    auto nodeIndex = startIndex;
-    auto nextNodeIndex = nodeIndex + 1;
-    Point p0 = m_nodes[nodeIndex];
-    Point p1 = m_nodes[nextNodeIndex];
-    double pointLengthCoordinate = nodeLengthCoordinate[startIndex];
-    bool snappedToLastPoint = false;
-    while (nodeIndex < endIndex)
+    for (size_t i = startIndex; i < endIndex; ++i)
     {
-        // initial point already accounted for
-        pointLengthCoordinate += refinementDistance;
-        if (pointLengthCoordinate > nodeLengthCoordinate[nextNodeIndex])
-        {
-            // if not snapped to the original last polygon point, snap it
-            if (!snappedToLastPoint)
-            {
-                refinedPolygon.emplace_back(m_nodes[nextNodeIndex]);
-            }
+        Point p = m_nodes[i];
+        Point delta = m_nodes[i + 1] - m_nodes[i];
+        double segmentLength = ComputeDistance(m_nodes[i], m_nodes[i + 1], m_projection);
+        double lengthAlongInterval = refinementDistance;
+        // TODO how large should the tolerance be? 1.0e-3 seems quite tight for this algorithm
+        double tolerance = 1.0e-3;
```

Comment:
Constants.hpp is a better suited place for constants. Make constexpr.

Reply:
Done